### PR TITLE
Version Update: kube client, openshift client, origin

### LINKED ISSUE #785: Support OpenShift > 3.6.0
We need to able to support newer versions of OpenShift.

Some other issues have already been open regarding 3.7.0 related bug.

This issue counts as a tracker for the main activity.

Ideally, to avoid breaking people workflow, we have to put together a PR that do:
- [x] upgrades OS to 3.7.0
- [x] upgrades Kubernetes Client to latest 3.x release
- [x] verifies Integration Test pass
- [x] manually verify that an integration can be manually created

### PR DESCRIPTION
**Pull request by paoloantinori**
fixes #785, #760 

IT tests and manually deploying an integration worked fine here.

Notice that now BuildConfig logs are bound to the java logging level set for `io.syndesis.openshift`

You have many ways to enable it, I do it patching directly the DeploymentConfig (since it's harder in the template, being the value inside a list):

```bash
oc env dc/syndesis-rest JAVA_OPTIONS="$(oc env dc/syndesis-rest --list |  grep -P '(?<=JAVA_OPTIONS=)(.*)' -o --color=never) -Dlogging.level.io.syndesis=DEBUG"
```

### CODE REVIEW

**Comment by jimmidyson** [comment]
Labelled this as WIP and for TP4 so we don't accidentally merge this for TP3 until we've had a deeper look and tested properly. @paoloantinori please let me know if that is a problem and this is actually required for TP3.

**Comment by paoloantinori** [comment]
Technically it's blocking the Support Page functionality, but I can still work around it and delay the complete integration until this is ready to be merged.

I agree it'd be better to a couple of people to validate it's not introducing any issue with their workflow.

Maybe @dsimansk  and someone representing the UI team

**Comment by paoloantinori** [comment]
@rhuss  is tp3 officially closed? can we merge this?

**Comment by oscerd** [comment]
It's not closed yet I guess..

**Comment by paoloantinori** [comment]
I've just rebased after Luca rename. @rhuss  can you give it a look?

**Comment by rhuss** [comment]
@paoloantinori yeah, will do. gimme a bit, still fighting with TP3 releasing ....

File: app/rest/openshift/src/main/java/io/syndesis/openshift/OpenShiftServiceImpl.java [hunk 1]
```diff
@@ -47,45 +55,20 @@ public void build(String name, DeploymentData deploymentData, InputStream tarInp
...
-        updateImageName(sName);
         ensureSecret(sName, deploymentData);
 
         openShiftClient.deploymentConfigs().withName(sName).deployLatest();
     }
 
-    // Only required temporarily while image triggers do not work
-    // We are updating the image name with the latest image reference before increasing
-    // the latestVersion number
-    private void updateImageName(String sName) {
```
Comment (rhuss) [inline app/rest/openshift/src/main/java/io/syndesis/openshift/OpenShiftServiceImpl.java]: Are image triggers for now without problems on 3.7 ? (and maybe later 3.6 ?)

File: app/rest/openshift/src/main/java/io/syndesis/openshift/OpenShiftServiceImpl.java [hunk 2]
```diff
@@ -193,7 +179,9 @@ private void ensureDeploymentConfig(String name, DeploymentData deploymentData)
             .withNewMetadata().addToLabels("integration", name).endMetadata()
             .withNewSpec()
             .addNewContainer()
-            .withImage(" ").withImagePullPolicy("Always").withName(name)
+                .withImage("")
```
Comment (rhuss) [inline app/rest/openshift/src/main/java/io/syndesis/openshift/OpenShiftServiceImpl.java]: At some point we had an issue with an 0 length string for the image name. Not sure anynmore where. Have you verified that this works ?

File: app/rest/openshift/src/main/java/io/syndesis/openshift/OpenShiftServiceImpl.java [hunk 3]
```diff
@@ -210,51 +198,60 @@ private void ensureDeploymentConfig(String name, DeploymentData deploymentData)
...
     }
 
     private void ensureBuildConfig(String name, DeploymentData deploymentData, String builderStreamTag, String imageStreamNamespace) {
+        LOG.debug("Create or Replace Build Config {}", name);
+
+        // propagate log level to the build pod
+        List<EnvVar> envVarList = new ArrayList<>();
+        envVarList.add(new EnvVar("MAVEN_OPTS", config.getMavenOptions(), null));
+        if(LOG.isDebugEnabled()){
+            envVarList.add(new EnvVar("BUILD_LOGLEVEL", "5", null));
```
Comment (rhuss) [inline app/rest/openshift/src/main/java/io/syndesis/openshift/OpenShiftServiceImpl.java]: Do we want to hardcode the loglevel here or is something worth to consider for configuration ? I think this is good enough for now.

**Comment by paoloantinori** [comment]
Hi @rhuss Unluckily no, **this PR is not compatible with os3.6.0**:

```
syndesis-rest-2-xvs3s syndesis-rest io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://openshift.default.svc/apis/build.openshift.io/v1/namespaces/myproject/builds?labelSelector=openshift.io/build-config.name%3Dpaolo&fieldSelector=status%3DRunning. Message: No field label conversion function found for version: build.openshift.io/v1. Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=No field label conversion function found for version: build.openshift.io/v1, metadata=ListMeta(resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).syndesis-rest-2-xvs3s syndesis-rest
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:470) ~[kubernetes-client-3.1.5.jar!/:na]syndesis-rest-2-xvs3s syndesis-rest
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:409) ~[kubernetes-client-3.1.5.jar!/:na]syndesis-rest-2-xvs3s syndesis-rest
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:379) ~[kubernetes-client-3.1.5.jar!/:na]syndesis-rest-2-xvs3s syndesis-rest
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:343) ~[kubernetes-client-3.1.5.jar!/:na]syndesis-rest-2-xvs3s syndesis-rest
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:327) ~[kubernetes-client-3.1.5.jar!/:na]syndesis-rest-2-xvs3s syndesis-rest
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:605) ~[kubernetes-client-3.1.5.jar!/:na]syndesis-rest-2-xvs3s syndesis-rest
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:70) ~[kubernetes-client-3.1.5.jar!/:na]syndesis-rest-2-xvs3s syndesis-rest
at io.syndesis.openshift.OpenShiftServiceImpl.isBuildStarted(OpenShiftServiceImpl.java:126) ~[openshift-1.2-SNAPSHOT.jar!/:1.2-SNAPSHOT]syndesis-rest-2-xvs3s syndesis-rest
at io.syndesis.controllers.integration.online.ActivateHandler.isBuildStarted(ActivateHandler.java:152) ~[controllers-1.2-SNAPSHOT.jar!/:1.2-SNAPSHOT]syndesis-rest-2-xvs3s syndesis-rest
at io.syndesis.controllers.integration.online.ActivateHandler.execute(ActivateHandler.java:94) ~[controllers-1.2-SNAPSHOT.jar!/:1.2-SNAPSHOT]syndesis-rest-2-xvs3s syndesis-rest
at io.syndesis.controllers.integration.IntegrationController.lambda$callStatusChangeHandler$10(IntegrationController.java:175) ~[controllers-1.2-SNAPSHOT.jar!/:1.2-SNAPSHOT]syndesis-rest-2-xvs3s syndesis-rest
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_141]syndesis-rest-2-xvs3s syndesis-rest
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_141]syndesis-rest-2-xvs3s syndesis-rest
at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_141]
```

**Comment by rhuss** [comment]
@paoloantinori and with

```
Server https://api.fuse-ignite.openshift.com:443
openshift v3.6.173.0.21
kubernetes v1.6.1+5115d708d7
```

??

**Comment by rhuss** [comment]
@paoloantinori How should we proceed here ? Actually all our real clusters on which we deploy are:

* Fuse Ignite Cluster: OpenShift Master:  v3.6.173.0.21 (online version 3.6.0.85)
* rh-idev cluster (staging): OpenShift Master: v3.6.173.0.49

Which means, as soon as we merge this PR, staging and the ignite cluster won't work anymore. As the UXD, Docs and UI people are relying on those installations to always reflect the current state, I still dare to make the update.

I think we need a clear strategy how we can jump for 3.6 to 3.7, not technical wise but process wise. Also, it looks for me like a bad idea to bring in 3.7 so late for testing, as I expect quite some work for th e upgrade (when following the discussions on the SME list about upgrades from 3.6. to 3.7)

@syndesisio/all @kcbabo @aileenc : Wdyt, how should we tackle this problem ?

**Comment by kcbabo** [comment]
Started a thread with OSO team to figure out what the version plans are for OSO Pro.  Main thing we need to make sure of is that we are aligned to version of OpenShift on Pro tier when we GA.

**Comment by paoloantinori** [comment]
Impacts on Support Page has been lifted, since we have took another approach there.

**Comment by chirino** [comment]
@paoloantinori can you stash this close this PR out but stash this branch, for when our clusters are ready to move to 3.7?

**Pull request closed without merging**